### PR TITLE
Refactor upgrade-charm

### PR DESCRIPTION
**Pull request by SimonRichardson**
## Description of change

The idea is that we can break the refreshing of a charm (previously
called upgrade charm) into distinct functions. Those functions can act
in a similar vein to the deploy command.

By passing in a larger config to each refreshing task, it can pick what
it requires in a smaller unit config and work on just that. This should
give rise to better unit test integration.

### Charmstore

```sh
$ juju bootstrap lxd test --no-gui
$ juju deploy cs:logstash-4
$ juju upgrade-charm logstash
Looking up metadata for charm cs:logstash (channel: stable)
Added charm "cs:logstash-5" to the model.
Leaving endpoints in "alpha": beat, client, elasticsearch, java
```

### Charmstore with no updates

```sh
$ juju bootstrap lxd test --no-gui
$ juju deploy cs:wordpress
$ juju upgrade-charm wordpress
ERROR already running latest charm "cs:wordpress-0"
```

### Local

```sh
$ juju bootstrap lxd test --no-gui
$ juju deploy ./testcharms/charm-repo/bionic/dummy
$ juju upgrade-charm dummy --path=./testcharms/charm-repo/bionic/dummy
Added charm "local:bionic/dummy-7" to the model.
```

### Switch

```sh
$ juju bootstrap lxd test --no-gui
$ juju deploy ./testcharms/charm-repo/bionic/dummy
$ juju upgrade-charm dummy --switch=cs:wordpress
Added charm "cs:wordpress-0" to the model.
Adding endpoint "cache" to default space "alpha"
Adding endpoint "db" to default space "alpha"
Adding endpoint "loadbalancer" to default space "alpha"
Adding endpoint "nfs" to default space "alpha"
Adding endpoint "website" to default space "alpha"
```

### Switch with force series

```sh
$ juju bootstrap lxd test --no-gui
$ juju deploy ./testcharms/charm-repo/bionic/dummy
$ juju upgrade-charm dummy --switch=cs:mysql --force-series
Looking up metadata for charm cs:mysql (channel: stable)
Added charm "cs:mysql-58" to the model.
Adding endpoint "ceph" to default space "alpha"
Adding endpoint "cluster" to default space "alpha"
Adding endpoint "data" to default space "alpha"
Adding endpoint "db-admin" to default space "alpha"
Adding endpoint "ha" to default space "alpha"
Adding endpoint "local-monitors" to default space "alpha"
Adding endpoint "master" to default space "alpha"
Adding endpoint "monitors" to default space "alpha"
Adding endpoint "munin" to default space "alpha"
Adding endpoint "nrpe-external-master" to default space "alpha"
Adding endpoint "shared-db" to default space "alpha"
Adding endpoint "slave" to default space "alpha"
Leaving endpoint in "alpha": db
```

### CODE REVIEW

File: cmd/juju/application/refresher/interface.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+import (
+	"github.com/juju/charm/v8"
+	"gopkg.in/macaroon.v2"
+
+	commoncharm "github.com/juju/juju/api/common/charm"
+	corecharm "github.com/juju/juju/core/charm"
+)
+
+// RefresherFactory contains a method to get a deployer.
```
Comment (hmlanigan) [inline cmd/juju/application/refresher/interface.go:14]: s/deployer/refresher

File: cmd/juju/application/refresher/interface.go [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+type RefresherFactory interface {
+	Run(RefresherConfig) (*CharmID, error)
+}
+
+// Refresher defines the functionality of a deployer returned by the
```
Comment (hmlanigan) [inline cmd/juju/application/refresher/interface.go:19]: s/deployer/refresher

File: cmd/juju/application/refresher/interface.go [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+// factory.
+type Refresher interface {
+	// Check will attempt to check if a refresher is allowed to run a given
```
Comment (hmlanigan) [inline cmd/juju/application/refresher/interface.go:22]: s/Check/Allowed/

File: cmd/juju/application/refresher/interface.go [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+	// config.
+	Allowed(RefresherConfig) (bool, error)
+	// Refresh finishes preparing to deploy a charm or bundle,
```
Comment (hmlanigan) [inline cmd/juju/application/refresher/interface.go:25]: Update comment for refresh.

File: cmd/juju/application/refresher/interface.go [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+}
+
+// CharmResolver defines methods required to resolve charms, as required
+// by the upgrade-charm command.
+type CharmResolver interface {
+	ResolveCharm(url *charm.URL, preferredOrigin commoncharm.Origin) (*charm.URL, commoncharm.Origin, []string, error)
+}
+
+// CharmRepo defines methods for interaction with a charm repo.
+type CharmRepo interface {
```
Comment (hmlanigan) [inline cmd/juju/application/refresher/interface.go:48]: Do we want to call this Repository?  To be the same as other such interfaces.

File: cmd/juju/application/refresher/refresher.go [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+		clock: jujuclock.WallClock,
+	}
+	d.refreshers = []RefresherFn{
+		d.maybeReadLocal(deps.CharmAdder, defaultCharmRepo{}),
+		d.maybeCharmStore(deps.Authorizer, deps.CharmAdder, deps.CharmResolver),
+	}
+	return d
+}
+
+// GetRefresher returns the correct deployer to use based on the cfg provided.
```
Comment (hmlanigan) [inline cmd/juju/application/refresher/refresher.go:65]: s/deployer/refresher

File: cmd/juju/application/refresher/refresher.go [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+// A ModelConfigGetter and CharmStoreAdaptor needed to find the deployer.
```
Comment (hmlanigan) [inline cmd/juju/application/refresher/refresher.go:66]: this comment line does not apply.

**Review by hmlanigan — APPROVED** [review]
LGTM for testing.  There are some comments to update.

**Comment by SimonRichardson** [comment]
$$merge$$

**Comment by SimonRichardson** [comment]
> FAIL: apiaddressupdater_test.go:107: APIAddressUpdaterSuite.TestAddressChange

$$merge$$

File: cmd/juju/application/refresher/refresher.go [hunk 8]
```diff
@@ -0,0 +1,268 @@
...
+		clock: jujuclock.WallClock,
+	}
+	d.refreshers = []RefresherFn{
+		d.maybeReadLocal(deps.CharmAdder, defaultCharmRepo{}),
+		d.maybeCharmStore(deps.Authorizer, deps.CharmAdder, deps.CharmResolver),
+	}
+	return d
+}
+
+// GetRefresher returns the correct refresher to use based on the cfg provided.
```
Comment (manadart) [inline cmd/juju/application/refresher/refresher.go:65]: Comment is wrong. There are some other exported methods in this patch missing comments too.

**Pull request merged**